### PR TITLE
Refactor effect inspection utils

### PR DESCRIPTION
**Pull request by shinima**
| Q                       | A              |
| ----------------------- | -------------- |
| Fixed Issues?           |                |
| Patch: Bug Fix?         |                |
| Major: Breaking Change? | Yes. See below |
| Minor: New Feature?     |                |
| Tests Added + Pass?     |                |
| Any Dependency Changes? |                |

API changes in this PR:

- API `asEffect` is removed
- API `is.effect` is added to test whether an object is an effect
- `effectTypes` is exported by `redux-saga/effects`.

I refactor the saga-monitor-example and split it into several modules. During refactoring the sagaMonitor, I find that there is an internal ROOT type. ROOT effects are used by sagaMonitor to indicate that a root saga starts to run. So I add ROOT as an effect type.

Effects inspection utils in this PR are relatively simple and straightforward. Documentation for effects should be updated and tells users that **Effect is an object that consists of a `type`, a `payload` and a internal symbol `IO`**. So users could import `is` and `effectTypes` to test whether an effect is of an specific type like in [checkers](https://github.com/redux-saga/redux-saga/pull/1555/files#diff-5d397e302f1805e01eb7aeb798443633).

How to inspect the details of the effect payload is still an question. This PR already uses the knowing of payload structure to log effects in a better format. We need to come up with a better way for users to access data in effect payload. Or maybe we can simply state in our doc: **You can log the effects and see the payload structure for debugging purposes, but you should not rely on it since payload structure may change over versions**.

### CODE REVIEW

**Comment by shinima** [comment]
By the way, saga-monitor remains the same style of logging output.

image

File: examples/sagaMonitor/modules/logSaga.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+  const winnerInd = desc.winner ? (isError ? '✘' : '✓') : ''
+  formatter.addLabel(winnerInd).addLabel(desc.label)
+
+  if (is.iterator(desc.effect)) {
+    formatter.addValue(desc.effect.name).addDescResult(desc, true)
+  } else if (is.promise(desc.effect)) {
+    formatter
+      .addEffectType('promise')
+      .resetStyle()
```
Comment (restrry) [inline examples/sagaMonitor/modules/logSaga.js]: why we need to call `resetStyle` manually for all the cases except `addLabel`?

File: examples/sagaMonitor/modules/DescriptorFormatter.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import Formatter from './Formatter'
+import { CANCELLED, PENDING, REJECTED, RESOLVED } from './constants'
+
+const DEFAULT_STYLE = 'color: black'
+const LABEL_STYLE = 'font-weight: bold'
+const EFFECT_TYPE_STYLE = 'color: blue'
+const ERROR_STYLE = 'color: red'
+const CANCEL_STYLE = 'color: #ccc'
+
+export default class DescriptorFormatter extends Formatter {
```
Comment (restrry) [inline examples/sagaMonitor/modules/DescriptorFormatter.js]: are we going to reuse `Formatter`? imho implementation with inheritance a bit confusing.

File: examples/sagaMonitor/modules/logSaga.js (hunk 1 above)
Comment (shinima) [reply examples/sagaMonitor/modules/logSaga.js]: This is adapted from the old code... Let me find a screenshot for you.

File: examples/sagaMonitor/modules/DescriptorFormatter.js (hunk 2 above)
Comment (shinima) [reply examples/sagaMonitor/modules/DescriptorFormatter.js]: The `Formatter` is a universal formatter which is adapted from the old code. And `DescriptorFormatter` is a special formatter that knows the structure of effect descriptors. So conceptually, `DescriptorFormatter` is a special kind of `Formatter`, so I use inheritance.

File: examples/sagaMonitor/modules/Manager.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  getRootIds() {
+    return this.rootIds
+  }
+
+  getChildIds(parentEffectId) {
+    return Object.keys(this.map)
+      .filter(effectId => this.map[effectId].parentEffectId === parentEffectId)
+      .map(Number)
```
Comment (Andarist) [inline examples/sagaMonitor/modules/Manager.js]: is this necessary? effect ids should be numbers already, right?

File: examples/sagaMonitor/modules/Manager.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  setAsRoot(effectId) {
+    this.rootIds.push(effectId)
+  }
+
+  getRootIds() {
+    return this.rootIds
+  }
+
+  getChildIds(parentEffectId) {
```
Comment (Andarist) [inline examples/sagaMonitor/modules/Manager.js]: seems like we should bookkeep the tree structure here too for perf reasons

File: examples/sagaMonitor/modules/Manager.js (hunk 3 above)
Comment (shinima) [reply examples/sagaMonitor/modules/Manager.js]: Keys of object are always strings, and `Object.keys(obj)` returns an array of strings.
image

File: examples/sagaMonitor/modules/Manager.js (hunk 4 above)
Comment (shinima) [reply examples/sagaMonitor/modules/Manager.js]: Yes. Keep an map from parent to children would enhance the performance.

**Comment by shinima** [comment]
Thanks for reviewing. And I'm currently thinking about add another package `@redux-saga/inpection` that exports some utils to allow users to inspect the internals of effects.  `@redux-saga/inpection` is for debugging purposes and `@redux-saga/core` does not depend on it.

I'll update this PR after implementing the `@redux-saga/inpection`.

File: packages/core/src/internal/io.js [hunk 5]
```diff
@@ -7,16 +7,19 @@ const TEST_HINT =
 
 const makeEffect = (type, payload) => ({ [IO]: true, type, payload })
 
-const isForkEffect = eff => eff && eff[IO] && eff.type === 'FORK'
-
 export const detach = eff => {
   if (process.env.NODE_ENV === 'development') {
+    const isForkEffect = eff => is.effect(eff) && eff.type === effectTypes.FORK
```
Comment (Andarist) [inline packages/core/src/internal/io.js]: why the change? why to move it to the function?

File: examples/sagaMonitor/modules/Manager.js (hunk 3 above)
Comment (Andarist) [reply examples/sagaMonitor/modules/Manager.js]: oh, yeah - havent quite noticed they are coming here from Object.keys

**Comment by Andarist** [comment]
Could u rebase this PR? 

I'm wondering if we need more inspection utils - they will only abstract simple object property access. I would be good with providing only `is.effect` + effectTypes. WDYT?

File: packages/core/src/internal/effectTypes.js [hunk 6]
```diff
@@ -1,3 +1,4 @@
+export const ROOT = 'ROOT'
```
Comment (Andarist) [inline packages/core/src/internal/effectTypes.js]: i dont think I quite like this change - seems like `ROOT` is different from other effects because its purely internal

could u describe what do u think about it? what are advantages of having it in your opinion? (remember that I can always be persuaded! 😉 )

File: packages/core/src/internal/io.js (hunk 5 above)
Comment (shinima) [reply packages/core/src/internal/io.js]: `isForkEffect` is used only in development, so I move it inside the if-block to get better code elimination. 

However it turns out that even if `isForkEffect` is placed at top level, the code of `isForkEffect` would be removed when redux-saga builds. 😅

File: packages/core/src/internal/effectTypes.js (hunk 6 above)
Comment (shinima) [reply packages/core/src/internal/effectTypes.js]: From the perspective of the people who uses redux-saga, ROOT is a different and should be kept internal only; From the perspective of saga-monitor, ROOT is of no difference compared to ALL/RACE/CALL.

IMO, ROOT is a valid effect type, but it does not need to be exported because users never yield a root effect directly. I'll remove it from *effectTypes.js*.

**Comment by shinima** [comment]
> I'm wondering if we need more inspection utils - they will only abstract simple object property access. I would be good with providing only is.effect + effectTypes. WDYT?

Actually I think **only is.effect + effectTypes** is fine. And as the effect payload shape is stable, I'm inclining to let people inspect the payload directly for (and only for) debugging purposes (we may add some tips into our docs about what an effect looks like).

**Comment by shinima** [comment]
Rebased on master. And add `manager.childIdsMap` for bookkeeping a map from parent-id to child-id-array, also removes ROOT from effectTypes.js.

File: packages/core/src/internal/io.js (hunk 5 above)
Comment (Andarist) [reply packages/core/src/internal/io.js]: yeah, in general function declarations and expressions are one of the easiest things to dead code eliminate, it's better to hoist this to the top level scope to avoid recreating this function inside each detach call

File: packages/core/src/internal/effectTypes.js (hunk 6 above)
Comment (Andarist) [reply packages/core/src/internal/effectTypes.js]: if we decide to keep it then it should stay in effectTypes, it shouldnt be exported from effects (as in `import {root} from 'redux-saga/effects'`

there is also a perspective of redux-saga itself - this effect is not interpreted like other effects, its executed eagerly on the startup

Comment (shinima) [reply packages/core/src/internal/effectTypes.js]: Actually users can not do `import {root} from 'redux-saga/effects'` as `root` is not re-exported in core/src/effects.js.

I'm a little bit of confused about what to do now... Should we add ROOT as a normal effect like ALL/RACE? Or should we remove ROOT from effectTypes, and keep `root` as internal? Or do you have any nicer ideas?

**Comment by shinima** [comment]
> it's better to hoist this to the top level scope to avoid recreating this function inside each detach call

done.

File: packages/core/src/internal/effectTypes.js (hunk 6 above)
Comment (Andarist) [reply packages/core/src/internal/effectTypes.js]: im quite sure it shouldnt be exported, im still unsure if it should be an effect at all 🤔 this only seems to satisfy monitor needs, from core's perspective it still seems to be weird imho - @restrry  wdyt?

Comment (Andarist) [reply packages/core/src/internal/effectTypes.js]: OTOH it is super minor thing, and was classified as an effect already when passing info about it to the monitor

Comment (restrry) [reply packages/core/src/internal/effectTypes.js]: imho that's not an effect per se, as it  has other lifecycle for end user (import - call - processing - return a result) and more over, users don't know about it, as it internal detail.

Comment (Andarist) [reply packages/core/src/internal/effectTypes.js]: it could be kept as purely internal effect - not user facing, but I think I like `rootSagaStarted` hook proposal by @shinima better, it's more explicit and less confusing about what effects are

Comment (restrry) [reply packages/core/src/internal/effectTypes.js]: +1

**Comment by shinima** [comment]
`rootSagaStarted` hook added.

File: packages/core/effects.d.ts [hunk 7]
```diff
@@ -706,3 +706,21 @@ export function debounce<T, T1, T2, T3, T4, T5, T6>(
   arg1: T1, arg2: T2, arg3: T3, arg4: T4, arg5: T5, arg6: T6,
   ...rest: any[]
 ): ForkEffect;
+
+export const effectTypes: {
```
Comment (Andarist) [inline packages/core/effects.d.ts]: I'd probably like to move it to separate module - but gonna handle it in separate PR

File: packages/core/src/utils.js [hunk 8]
```diff
@@ -1,4 +1,2 @@
 export { SAGA_ACTION, TASK } from '@redux-saga/symbols'
```
Comment (Andarist) [inline packages/core/src/utils.js]: the `redux-saga/utils` entry is starting to look useless, we should get rid of it later in favour of modular packages
